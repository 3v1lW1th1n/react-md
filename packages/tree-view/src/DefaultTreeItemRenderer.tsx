import * as React from "react";
import { IListItemBaseProps } from "@react-md/list";

import { ITreeViewItemInjectedProps } from "./types";
import { default as TreeItem, ITreeItemBaseProps } from "./TreeItem";
import TreeLinkItem from "./TreeLinkItem";
import TreeItemContent from "./TreeItemContent";
import TreeItemExpanderIcon from "./TreeItemExpanderIcon";
import TreeGroup from "./TreeGroup";

export interface IDefaultTreeItemRendererBaseProps extends IListItemBaseProps {
  [key: string]: any;
  href?: string;
  expanderLeft?: boolean;
  linkComponent?: React.ReactType;
}

interface IA11yProps {
  "aria-expanded"?: "true";
  "aria-level": number;
  "aria-setsize": number;
  "aria-posinset": number;
  tabIndex: 0 | -1;
}

export type IDefaultTreeItemRendererProps = IDefaultTreeItemRendererBaseProps & ITreeViewItemInjectedProps;

export interface IDefaultTreeItemRendererState {}

export default class DefaultTreeItemRenderer extends React.Component<
  IDefaultTreeItemRendererProps,
  IDefaultTreeItemRendererState
> {
  constructor(props: IDefaultTreeItemRendererProps) {
    super(props);

    this.state = {};
  }

  public componentDidMount() {
    this.props.updateTreeItems();
  }

  public componentWillUnmount() {
    this.props.updateTreeItems();
  }

  public render() {
    const {
      depth,
      listSize,
      itemIndex,
      expanderLeft,
      linkComponent: propLinkComponent,
      updateTreeItems,
      renderChildItems,
      expanded,
      children,
      forceIconWrap: propForceIconWrap,
      leftIcon: propLeftIcon,
      rightIcon: propRightIcon,
      ...props
    } = this.props;
    const linkComponent = propLinkComponent || (typeof props.href === "string" && "a") || undefined;
    const forceIconWrap = propForceIconWrap || !!renderChildItems;
    const { selected } = props;
    let group;
    let leftIcon = propLeftIcon;
    let rightIcon = propRightIcon;
    if (renderChildItems) {
      const icon = <TreeItemExpanderIcon rotated={expanded} />;
      if (expanderLeft) {
        leftIcon = icon;
      } else {
        rightIcon = icon;
      }

      group = <TreeGroup expanded={expanded}>{renderChildItems()}</TreeGroup>;
    }

    const a11y = {
      "aria-expanded": expanded ? "true" : undefined,
      "aria-level": depth + 1,
      "aria-setsize": listSize,
      "aria-posinset": itemIndex + 1,
      tabIndex: selected ? 0 : -1,
    } as IA11yProps;

    if (linkComponent) {
      return (
        <TreeLinkItem
          {...props}
          {...a11y}
          linkComponent={linkComponent}
          leftIcon={leftIcon}
          rightIcon={rightIcon}
          forceIconWrap={forceIconWrap}
          group={group}
        >
          {children}
        </TreeLinkItem>
      );
    }

    return (
      <TreeItem {...props} {...a11y}>
        <TreeItemContent forceIconWrap={forceIconWrap} leftIcon={leftIcon} rightIcon={rightIcon} selected={selected}>
          {children}
        </TreeItemContent>
        {group}
      </TreeItem>
    );
  }
}
