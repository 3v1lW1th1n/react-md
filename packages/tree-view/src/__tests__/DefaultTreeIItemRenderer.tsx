import * as React from "react";
import * as renderer from "react-test-renderer";

import DefaultTreeItemRenderer from "../DefaultTreeItemRenderer";

const BASE_PROPS = {
  depth: 1,
  listSize: 1,
  itemIndex: 1,
  selected: false,
  expanded: false,
  updateTreeItems: () => undefined,
};

// TODO: write more tests when enzyme supports new React context api
describe("DefaultTreeItemRenderer", () => {
  it("should render correctly with the provided props", () => {
    expect(
      renderer.create(<DefaultTreeItemRenderer {...BASE_PROPS}>Content!</DefaultTreeItemRenderer>).toJSON()
    ).toMatchSnapshot();

    expect(
      renderer
        .create(
          <DefaultTreeItemRenderer {...BASE_PROPS} expanded={true}>
            Content!
          </DefaultTreeItemRenderer>
        )
        .toJSON()
    ).toMatchSnapshot();

    expect(
      renderer
        .create(
          <DefaultTreeItemRenderer {...BASE_PROPS} selected={true}>
            Content!
          </DefaultTreeItemRenderer>
        )
        .toJSON()
    ).toMatchSnapshot();

    expect(
      renderer
        .create(
          <DefaultTreeItemRenderer {...BASE_PROPS} expanded={true} selected={true}>
            Content!
          </DefaultTreeItemRenderer>
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  it("should render correctly as a link with the provided props", () => {
    const props = { ...BASE_PROPS, href: "/" };
    expect(
      renderer.create(<DefaultTreeItemRenderer {...props}>Content!</DefaultTreeItemRenderer>).toJSON()
    ).toMatchSnapshot();

    expect(
      renderer
        .create(
          <DefaultTreeItemRenderer {...props} expanded={true}>
            Content!
          </DefaultTreeItemRenderer>
        )
        .toJSON()
    ).toMatchSnapshot();

    expect(
      renderer
        .create(
          <DefaultTreeItemRenderer {...props} selected={true}>
            Content!
          </DefaultTreeItemRenderer>
        )
        .toJSON()
    ).toMatchSnapshot();

    expect(
      renderer
        .create(
          <DefaultTreeItemRenderer {...props} expanded={true} selected={true}>
            Content!
          </DefaultTreeItemRenderer>
        )
        .toJSON()
    ).toMatchSnapshot();
  });
});
