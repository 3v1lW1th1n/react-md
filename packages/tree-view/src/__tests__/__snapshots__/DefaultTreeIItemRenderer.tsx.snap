// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultTreeItemRenderer should render correctly as a link with the provided props 1`] = `
<li
  className="rmd-tree-item"
  role="none"
>
  <a
    aria-level={1}
    aria-posinset={1}
    aria-setsize={1}
    className="rmd-tree-item__content rmd-tree-item__content--link rmd-list-item rmd-list-item--stateful rmd-states--simple"
    disabled={false}
    expanded={false}
    href="/"
    role="treeitem"
    tabIndex={-1}
    updateTreeItems={[Function]}
  >
    <span
      className="rmd-list-item__text"
    >
      Content!
    </span>
  </a>
</li>
`;

exports[`DefaultTreeItemRenderer should render correctly as a link with the provided props 2`] = `
<li
  className="rmd-tree-item"
  role="none"
>
  <a
    aria-expanded="true"
    aria-level={1}
    aria-posinset={1}
    aria-setsize={1}
    className="rmd-tree-item__content rmd-tree-item__content--link rmd-list-item rmd-list-item--stateful rmd-states--simple"
    disabled={false}
    expanded={true}
    href="/"
    role="treeitem"
    tabIndex={-1}
    updateTreeItems={[Function]}
  >
    <span
      className="rmd-list-item__text"
    >
      Content!
    </span>
  </a>
</li>
`;

exports[`DefaultTreeItemRenderer should render correctly as a link with the provided props 3`] = `
<li
  className="rmd-tree-item"
  role="none"
>
  <a
    aria-level={1}
    aria-posinset={1}
    aria-setsize={1}
    className="rmd-tree-item__content rmd-tree-item__content--link rmd-list-item rmd-list-item--stateful rmd-states--simple rmd-states--selected"
    disabled={false}
    expanded={false}
    href="/"
    role="treeitem"
    tabIndex={0}
    updateTreeItems={[Function]}
  >
    <span
      className="rmd-list-item__text"
    >
      Content!
    </span>
  </a>
</li>
`;

exports[`DefaultTreeItemRenderer should render correctly as a link with the provided props 4`] = `
<li
  className="rmd-tree-item"
  role="none"
>
  <a
    aria-expanded="true"
    aria-level={1}
    aria-posinset={1}
    aria-setsize={1}
    className="rmd-tree-item__content rmd-tree-item__content--link rmd-list-item rmd-list-item--stateful rmd-states--simple rmd-states--selected"
    disabled={false}
    expanded={true}
    href="/"
    role="treeitem"
    tabIndex={0}
    updateTreeItems={[Function]}
  >
    <span
      className="rmd-list-item__text"
    >
      Content!
    </span>
  </a>
</li>
`;

exports[`DefaultTreeItemRenderer should render correctly with the provided props 1`] = `
<li
  aria-level={1}
  aria-posinset={1}
  aria-setsize={1}
  className="rmd-tree-item rmd-states--simple"
  disabled={false}
  expanded={false}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  role="treeitem"
  tabIndex={-1}
  updateTreeItems={[Function]}
>
  <div
    className="rmd-tree-item__content rmd-list-item rmd-list-item--stateful"
  >
    <span
      className="rmd-list-item__text"
    >
      Content!
    </span>
  </div>
</li>
`;

exports[`DefaultTreeItemRenderer should render correctly with the provided props 2`] = `
<li
  aria-expanded="true"
  aria-level={1}
  aria-posinset={1}
  aria-setsize={1}
  className="rmd-tree-item rmd-states--simple"
  disabled={false}
  expanded={true}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  role="treeitem"
  tabIndex={-1}
  updateTreeItems={[Function]}
>
  <div
    className="rmd-tree-item__content rmd-list-item rmd-list-item--stateful"
  >
    <span
      className="rmd-list-item__text"
    >
      Content!
    </span>
  </div>
</li>
`;

exports[`DefaultTreeItemRenderer should render correctly with the provided props 3`] = `
<li
  aria-level={1}
  aria-posinset={1}
  aria-setsize={1}
  className="rmd-tree-item rmd-states--simple rmd-states--selected"
  disabled={false}
  expanded={false}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  role="treeitem"
  tabIndex={0}
  updateTreeItems={[Function]}
>
  <div
    className="rmd-tree-item__content rmd-list-item rmd-list-item--stateful"
  >
    <span
      className="rmd-list-item__text"
    >
      Content!
    </span>
  </div>
</li>
`;

exports[`DefaultTreeItemRenderer should render correctly with the provided props 4`] = `
<li
  aria-expanded="true"
  aria-level={1}
  aria-posinset={1}
  aria-setsize={1}
  className="rmd-tree-item rmd-states--simple rmd-states--selected"
  disabled={false}
  expanded={true}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  role="treeitem"
  tabIndex={0}
  updateTreeItems={[Function]}
>
  <div
    className="rmd-tree-item__content rmd-list-item rmd-list-item--stateful"
  >
    <span
      className="rmd-list-item__text"
    >
      Content!
    </span>
  </div>
</li>
`;
